src/vfs.rs

```diff
@@ -0,0 +1,146 @@
+use crate::common::*;
+
+#[derive(Debug, Clone)]
+pub(crate) struct Vfs {
+  base_directory: InputPath,
+}
+
+impl Vfs {
+  pub(crate) fn new(base_directory: InputPath) -> Self {
+    Self { base_directory }
+  }
+
+  /// If an `.index.md` file exists in this directory, return its contents as a string.
+  pub(crate) fn directory_markdown_file(&self, dir_path: &InputPath) -> Result<Option<String>> {
+    let file = dir_path.join_relative(".index.md".as_ref())?;
+    match fs::read_to_string(&file) {
+      Ok(markdown) => Ok(Some(markdown)),
+      Err(error) if error.kind() == io::ErrorKind::NotFound => Ok(None),
+      Err(source) => Err(Error::filesystem_io(&file).into_error(source)),
+    }
+  }
+
+  pub(crate) fn paid(&self, path: &InputPath) -> bool {
+    let config = Config::for_dir(
+      self.base_directory.as_ref(),
+      path.as_ref().parent().unwrap(),
+    )
+    .unwrap();
+    config.paid()
+  }
+
+  pub(crate) fn base_price(&self, path: &InputPath) -> Option<Millisatoshi> {
```

Comment:
This unwrap should be removed, `base_price` can return `Result<Option<Millisatoshi>>`.